libraries/AP_Proximity/AP_Proximity_LD06.cpp

```diff
@@ -0,0 +1,169 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * ArduPilot device driver for Inno-Maker LD06 LiDAR
+ *
+ * ALL INFORMATION REGARDING PROTOCOL WAS DERIVED FROM InnoMaker DATASHEET:
+ *
+ * http://wiki.inno-maker.com/display/HOMEPAGE/LD06?preview=/6949506/6949511/LDROBOT_LD06_Development%20manual_v1.0_en.pdf
+ *
+ * Author: Adithya Patil, Georgia Institute of Technology
+ * Based on the SLAMTEC RPLiDAR code written by Steven Josefs, IAV GmbH and CYGBOT D1 LiDAR code
+ *
+ */
+
+#include "AP_Proximity_LD06.h"
+
+#if HAL_PROXIMITY_ENABLED && AP_PROXIMITY_LD06_ENABLED
+
+// Update the sensor readings
+void AP_Proximity_LD06::update(void)
+{
+    // Escape if no connection detected/supported while running
+    if (_uart == nullptr) {
+        return;
+    }
+
+    // Begin getting sensor readings
+    // Calls method that repeatedly reads through UART channel
+    get_readings();
+
+    // Check if the data is being received correctly and sets Proximity Status
+    if (_last_distance_received_ms == 0 || (AP_HAL::millis() - _last_distance_received_ms > PROXIMITY_LD06_TIMEOUT_MS)) {
+        set_status(AP_Proximity::Status::NoData);
+    } else {
+        set_status(AP_Proximity::Status::Good);
+    }
+}
+
+// Called repeatedly to get the readings at the current instant
+void AP_Proximity_LD06::get_readings()
+{
+    if (_uart == nullptr) {
+        return;
+    }
+
+    // Store the number of bytes available on the UART input
+    uint32_t nbytes = MIN((uint16_t) 4000,  _uart->available());
+
+    // Loops through all bytes that were received
+    while (nbytes-- > 0) {
+
+        // Gets and logs the current byte being read
+        const uint8_t c = _uart->read();
+
+        // Stores the byte in an array if the byte is a start byte or we have already read a start byte
+        if (c == LD_START_CHAR || _response_data) {
+
+            // Sets to true if a start byte has been read, default false otherwise
+            _response_data = true;
+
+            // Stores the next byte in an array
+            _response[_byte_count] = c;
+            _byte_count++;
+
+            if (_byte_count == _response[START_DATA_LENGTH] + 3) {
+
+                set_status(AP_Proximity::Status::Good);
+
+                uint32_t current_ms = AP_HAL::millis();
```

Comment:
`const uint32_t`

Reply:
@rishabsingh3003 could you commit the required changes? Would be the quickest way to finish the work on this driver?

Reply:
@lvale I don't think I have the "maintainer" access to commit the required changes in this PR. I could open another PR with the suggested fixes, but I would be more comfortable in doing that if the sensor were sent to me or if someone could do the simulator for this and confirm that it works. Or maybe someone can just send me a raw serial dump of the sensor and I could test if everything works. 
The changes I have asked are minimal though

Reply:
just give me a couple of days to sort things here and commit these changes and will send log with this model attached on serial

Reply:
working now on Rover, but can test on other vehicules. What kind of logging do you want?

![image](https://github.com/ArduPilot/ardupilot/assets/3130363/7ba5fbf6-fea5-4289-a913-e78c3841ba39)

Reply:
I have a log
attaching here
[log_0_UnknownDate.bin.zip](https://github.com/ArduPilot/ardupilot/files/12241700/log_0_UnknownDate.bin.zip)

@tridge as requested.

Distances correspond to observed physical distances, within just a few centimeters margin. Quite surprised by the lidar accuracy.